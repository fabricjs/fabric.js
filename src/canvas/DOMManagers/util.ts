import { LEFT, NONE, TOP } from '../../constants';
import type { TSize } from '../../typedefs';
import {
  getDocumentFromElement,
  getWindowFromElement,
  getScrollLeftTop,
} from '../../util/dom_misc';
import { setStyle } from '../../util/dom_style';

export const setCanvasDimensions = (
  el: HTMLCanvasElement,
  ctx: CanvasRenderingContext2D,
  { width, height }: TSize,
  retinaScaling = 1
) => {
  el.width = width;
  el.height = height;
  if (retinaScaling > 1) {
    el.setAttribute('width', (width * retinaScaling).toString());
    el.setAttribute('height', (height * retinaScaling).toString());
    ctx.scale(retinaScaling, retinaScaling);
  }
};

export function allowTouchScrolling(element: HTMLElement, allow: boolean) {
  const touchAction = allow ? 'manipulation' : NONE;
  setStyle(element, {
    'touch-action': touchAction,
    '-ms-touch-action': touchAction,
  });
}

export type CSSDimensions = {
  width: number | `${number}${string}`;
  height: number | `${number}${string}`;
};

export const setCSSDimensions = (
  el: HTMLElement,
  { width, height }: Partial<CSSDimensions>
) => {
  width && (el.style.width = typeof width === 'number' ? `${width}px` : width);
  height &&
    (el.style.height = typeof height === 'number' ? `${height}px` : height);
};

/**
 * Returns offset for a given element
 * @param {HTMLElement} element Element to get offset for
 * @return {Object} Object with "left" and "top" properties
 */
export function getElementOffset(element: HTMLElement) {
  let box = { left: 0, top: 0 };
  const doc = element && getDocumentFromElement(element),
    offset = { left: 0, top: 0 },
    offsetAttributes = {
      borderLeftWidth: LEFT,
      borderTopWidth: TOP,
      paddingLeft: LEFT,
      paddingTop: TOP,
    } as const;

  if (!doc) {
    return offset;
  }
  const elemStyle =
    getWindowFromElement(element)?.getComputedStyle(element, null) || {};
  for (const attr in offsetAttributes) {
    // @ts-expect-error TS learn to iterate!
    offset[offsetAttributes[attr]] += parseInt(elemStyle[attr], 10) || 0;
  }

  const docElem = doc.documentElement;
  if (typeof element.getBoundingClientRect !== 'undefined') {
    box = element.getBoundingClientRect();
  }

  const scrollLeftTop = getScrollLeftTop(element);

  return {
    left:
      box.left + scrollLeftTop.left - (docElem.clientLeft || 0) + offset.left,
    top: box.top + scrollLeftTop.top - (docElem.clientTop || 0) + offset.top,
  };
}

/**
 * Makes element unselectable
 * @param {HTMLElement} element Element to make unselectable
 * @return {HTMLElement} Element that was passed in
 */
export function makeElementUnselectable(element: HTMLElement) {
  if (typeof element.onselectstart !== 'undefined') {
    element.onselectstart = () => false;
  }
  element.style.userSelect = NONE;
  return element;
}
