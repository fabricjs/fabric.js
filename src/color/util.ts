/**
 * @param {Number} p
 * @param {Number} q
 * @param {Number} t
 * @return {Number}
 */
export function hue2rgb(p: number, q: number, t: number): number {
  if (t < 0) {
    t += 1;
  }
  if (t > 1) {
    t -= 1;
  }
  if (t < 1 / 6) {
    return p + (q - p) * 6 * t;
  }
  if (t < 1 / 2) {
    return q;
  }
  if (t < 2 / 3) {
    return p + (q - p) * (2 / 3 - t) * 6;
  }
  return p;
}

/**
 * Convert a value âˆˆ [0, 255] to hex
 */
export function hexify(value: number) {
  const hexValue = value.toString(16).toUpperCase();
  return hexValue.length === 1 ? `0${hexValue}` : hexValue;
}
