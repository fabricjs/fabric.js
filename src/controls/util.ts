import {
  TPointerEvent,
  Transform,
  TransformAction,
  BasicTransformEvent,
} from '../EventTypeDefs';
import { resolveOrigin } from '../util/misc/resolveOrigin';
import { Point } from '../Point';
import type { FabricObject } from '../shapes/Object/FabricObject';
import { TOriginX, TOriginY } from '../typedefs';
import {
  degreesToRadians,
  radiansToDegrees,
} from '../util/misc/radiansDegreesConversion';
import type { Control } from './Control';

export const NOT_ALLOWED_CURSOR = 'not-allowed';

/**
 * @param {Boolean} alreadySelected true if target is already selected
 * @param {String} corner a string representing the corner ml, mr, tl ...
 * @param {Event} e Event object
 * @param {FabricObject} [target] inserted back to help overriding. Unused
 */
export const getActionFromCorner = (
  alreadySelected: boolean,
  corner: string,
  e: TPointerEvent,
  target: FabricObject
) => {
  if (!corner || !alreadySelected) {
    return 'drag';
  }
  const control = target.controls[corner];
  return control.getActionName(e, control, target);
};

/**
 * Checks if transform is centered
 * @param {Object} transform transform data
 * @return {Boolean} true if transform is centered
 */
export function isTransformCentered(transform: Transform) {
  return transform.originX === 'center' && transform.originY === 'center';
}

export function invertOrigin(origin: TOriginX | TOriginY) {
  return -resolveOrigin(origin) + 0.5;
}

export const isLocked = (
  target: FabricObject,
  lockingKey:
    | 'lockMovementX'
    | 'lockMovementY'
    | 'lockRotation'
    | 'lockScalingX'
    | 'lockScalingY'
    | 'lockSkewingX'
    | 'lockSkewingY'
    | 'lockScalingFlip'
) => target[lockingKey];

export const commonEventInfo: TransformAction<
  Transform,
  BasicTransformEvent
> = (eventData, transform, x, y) => {
  return {
    e: eventData,
    transform,
    pointer: new Point(x, y),
  };
};

/**
 * Combine control position and object angle to find the control direction compared
 * to the object center.
 * @param {FabricObject} fabricObject the fabric object for which we are rendering controls
 * @param {Control} control the control class
 * @return {Number} 0 - 7 a quadrant number
 */
export function findCornerQuadrant(
  fabricObject: FabricObject,
  control: Control
) {
  //  angle is relative to canvas plane
  const angle = fabricObject.getTotalAngle(),
    cornerAngle =
      angle + radiansToDegrees(Math.atan2(control.y, control.x)) + 360;
  return Math.round((cornerAngle % 360) / 45);
}

/**
 * @returns the normalized point (rotated relative to center) in local coordinates
 */
function normalizePoint(
  target: FabricObject,
  point: Point,
  originX: TOriginX,
  originY: TOriginY
): Point {
  const center = target.getRelativeCenterPoint(),
    p =
      typeof originX !== 'undefined' && typeof originY !== 'undefined'
        ? target.translateToGivenOrigin(
            center,
            'center',
            'center',
            originX,
            originY
          )
        : new Point(target.left, target.top),
    p2 = target.angle
      ? point.rotate(-degreesToRadians(target.angle), center)
      : point;
  return p2.subtract(p);
}

/**
 * Transforms a point to the offset from the given origin
 * @param {Object} transform
 * @param {String} originX
 * @param {String} originY
 * @param {number} x
 * @param {number} y
 * @return {Fabric.Point} the normalized point
 */
export function getLocalPoint(
  { target, corner }: Transform,
  originX: TOriginX,
  originY: TOriginY,
  x: number,
  y: number
) {
  const control = target.controls[corner],
    zoom = target.canvas?.getZoom() || 1,
    padding = target.padding / zoom,
    localPoint = normalizePoint(target, new Point(x, y), originX, originY);
  if (localPoint.x >= padding) {
    localPoint.x -= padding;
  }
  if (localPoint.x <= -padding) {
    localPoint.x += padding;
  }
  if (localPoint.y >= padding) {
    localPoint.y -= padding;
  }
  if (localPoint.y <= padding) {
    localPoint.y += padding;
  }
  localPoint.x -= control.offsetX;
  localPoint.y -= control.offsetY;
  return localPoint;
}
