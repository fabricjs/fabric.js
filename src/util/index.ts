export { cos } from './misc/cos';
export { sin } from './misc/sin';
export {
  rotateVector,
  createVector,
  calcAngleBetweenVectors,
  getUnitVector,
  calcVectorRotation,
  crossProduct,
  dotProduct,
  getOrthonormalVector,
  isBetweenVectors,
  magnitude,
} from './misc/vectors';
export {
  degreesToRadians,
  radiansToDegrees,
} from './misc/radiansDegreesConversion';
export { rotatePoint } from './misc/rotatePoint';
export * from './misc/projectStroke';
export type {
  TComposeMatrixArgs,
  TQrDecomposeOut,
  TRotateMatrixArgs,
  TScaleMatrixArgs,
  TTranslateMatrixArgs,
} from './misc/matrix';
export {
  transformPoint,
  invertTransform,
  composeMatrix,
  qrDecompose,
  createTranslateMatrix,
  createRotateMatrix,
  createScaleMatrix,
  createSkewXMatrix,
  createSkewYMatrix,
  calcDimensionsMatrix,
  multiplyTransformMatrices,
  multiplyTransformMatrixArray,
  isIdentityMatrix,
} from './misc/matrix';
export type { TextStyleArray } from './misc/textStyles';
export {
  stylesFromArray,
  stylesToArray,
  hasStyleChanged,
} from './misc/textStyles';
export {
  createCanvasElement,
  createImage,
  copyCanvasElement,
  toDataURL,
} from './misc/dom';
export { toFixed } from './misc/toFixed';
export {
  matrixToSVG,
  parsePreserveAspectRatioAttribute,
  parseUnit,
  getSvgAttributes,
} from './misc/svgParsing';
export { groupSVGElements } from './misc/groupSVGElements';
export { findScaleToFit, findScaleToCover } from './misc/findScaleTo';
export { capValue } from './misc/capValue';
export {
  saveObjectTransform,
  resetObjectTransform,
  addTransformToObject,
  applyTransformToObject,
  removeTransformFromObject,
  sizeAfterTransform,
} from './misc/objectTransforms';
export { makeBoundingBoxFromPoints } from './misc/boundingBoxFromPoints';
export {
  calcPlaneChangeMatrix,
  sendPointToPlane,
  sendVectorToPlane,
  sendObjectToPlane,
} from './misc/planeChange';
export * as string from './lang_string';
export type {
  EnlivenObjectOptions,
  LoadImageOptions,
} from './misc/objectEnlive';
export {
  loadImage,
  enlivenObjects,
  enlivenObjectEnlivables,
} from './misc/objectEnlive';
export { pick } from './misc/pick';
export * from './path/typedefs';
export {
  joinPath,
  parsePath,
  makePathSimpler,
  getSmoothPathFromPoints,
  getPathSegmentsInfo,
  getBoundsOfCurve,
  getPointOnPath,
  transformPath,
  getRegularPolygonPath,
} from './path';
export { setStyle } from './dom_style';
export { isTouchEvent, getPointer } from './dom_event';
export { isTransparent } from './misc/isTransparent';
export { mergeClipPaths } from './misc/mergeClipPaths';
export * from './animation';
export * as ease from './animation/easing';
export {
  requestAnimFrame,
  cancelAnimFrame,
} from './animation/AnimationFrameProvider';
export { removeFromArray } from './internals/removeFromArray';
export { getRandomInt } from './internals/getRandomInt';
export { request } from './internals/dom_request';

// for test compatibility. We don't want to export it
export { removeTransformMatrixForSvgParsing } from './transform_matrix_removal';
