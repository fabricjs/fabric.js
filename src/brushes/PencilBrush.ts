import { ModifierKey, TEvent } from '../EventTypeDefs';
import { Point } from '../Point';
import { Shadow } from '../Shadow';
import { Path } from '../shapes/Path';
import { PathData } from '../typedefs';
import { getSmoothPathFromPoints, joinPath } from '../util/path';
import type { Canvas } from '../canvas/Canvas';
import { BaseBrush } from './BaseBrush';

/**
 * @private
 * @param {PathData} pathData SVG path commands
 * @returns {boolean}
 */
function isEmptySVGPath(pathData: PathData): boolean {
  return joinPath(pathData) === 'M 0 0 Q 0 0 0 0 L 0 0';
}

export class PencilBrush extends BaseBrush {
  /**
   * Discard points that are less than `decimate` pixel distant from each other
   * @type Number
   * @default 0.4
   */
  decimate = 0.4;

  /**
   * Draws a straight line between last recorded point to current pointer
   * Used for `shift` functionality
   *
   * @type boolean
   * @default false
   */
  drawStraightLine = false;

  /**
   * The event modifier key that makes the brush draw a straight line.
   * If `null` or 'none' or any other string that is not a modifier key the feature is disabled.
   * @type {ModifierKey | undefined | null}
   */
  straightLineKey: ModifierKey | undefined | null = 'shiftKey';

  private declare _points: Point[];
  private declare _hasStraightLine: boolean;
  private declare oldEnd?: Point;

  constructor(canvas: Canvas) {
    super(canvas);
    this._points = [];
    this._hasStraightLine = false;
  }

  needsFullRender() {
    return super.needsFullRender() || this._hasStraightLine;
  }

  static drawSegment(ctx: CanvasRenderingContext2D, p1: Point, p2: Point) {
    const midPoint = p1.midPointFrom(p2);
    ctx.quadraticCurveTo(p1.x, p1.y, midPoint.x, midPoint.y);
    return midPoint;
  }

  /**
   * Invoked on mouse down
   * @param {Point} pointer
   */
  onMouseDown(pointer: Point, { e }: TEvent) {
    if (!this.canvas._isMainEvent(e)) {
      return;
    }
    this.drawStraightLine = !!this.straightLineKey && e[this.straightLineKey];
    this._prepareForDrawing(pointer);
    // capture coordinates immediately
    // this allows to draw dots (when movement never occurs)
    this._addPoint(pointer);
    this._render();
  }

  /**
   * Invoked on mouse move
   * @param {Point} pointer
   */
  onMouseMove(pointer: Point, { e }: TEvent) {
    if (!this.canvas._isMainEvent(e)) {
      return;
    }
    this.drawStraightLine = !!this.straightLineKey && e[this.straightLineKey];
    if (this.limitedToCanvasSize === true && this._isOutSideCanvas(pointer)) {
      return;
    }
    if (this._addPoint(pointer) && this._points.length > 1) {
      if (this.needsFullRender()) {
        // redraw curve
        // clear top canvas
        this.canvas.clearContext(this.canvas.contextTop);
        this._render();
      } else {
        const points = this._points,
          length = points.length,
          ctx = this.canvas.contextTop;
        // draw the curve update
        this._saveAndTransform(ctx);
        if (this.oldEnd) {
          ctx.beginPath();
          ctx.moveTo(this.oldEnd.x, this.oldEnd.y);
        }
        this.oldEnd = PencilBrush.drawSegment(
          ctx,
          points[length - 2],
          points[length - 1]
        );
        ctx.stroke();
        ctx.restore();
      }
    }
  }

  /**
   * Invoked on mouse up
   */
  onMouseUp({ e }: TEvent) {
    if (!this.canvas._isMainEvent(e)) {
      return true;
    }
    this.drawStraightLine = false;
    this.oldEnd = undefined;
    this._finalizeAndAddPath();
    return false;
  }

  /**
   * @private
   * @param {Point} pointer Actual mouse position related to the canvas.
   */
  _prepareForDrawing(pointer: Point) {
    this._reset();
    this._addPoint(pointer);
    this.canvas.contextTop.moveTo(pointer.x, pointer.y);
  }

  /**
   * @private
   * @param {Point} point Point to be added to points array
   */
  _addPoint(point: Point) {
    if (
      this._points.length > 1 &&
      point.eq(this._points[this._points.length - 1])
    ) {
      return false;
    }
    if (this.drawStraightLine && this._points.length > 1) {
      this._hasStraightLine = true;
      this._points.pop();
    }
    this._points.push(point);
    return true;
  }

  /**
   * Clear points array and set contextTop canvas style.
   * @private
   */
  _reset() {
    this._points = [];
    this._setBrushStyles(this.canvas.contextTop);
    this._setShadow();
    this._hasStraightLine = false;
  }

  /**
   * Draw a smooth path on the topCanvas using quadraticCurveTo
   * @private
   * @param {CanvasRenderingContext2D} [ctx]
   */
  _render(ctx: CanvasRenderingContext2D = this.canvas.contextTop) {
    let p1 = this._points[0],
      p2 = this._points[1];
    this._saveAndTransform(ctx);
    ctx.beginPath();
    //if we only have 2 points in the path and they are the same
    //it means that the user only clicked the canvas without moving the mouse
    //then we should be drawing a dot. A path isn't drawn between two identical dots
    //that's why we set them apart a bit
    if (this._points.length === 2 && p1.x === p2.x && p1.y === p2.y) {
      const width = this.width / 1000;
      p1.x -= width;
      p2.x += width;
    }
    ctx.moveTo(p1.x, p1.y);

    for (let i = 1; i < this._points.length; i++) {
      // we pick the point between pi + 1 & pi + 2 as the
      // end point and p1 as our control point.
      PencilBrush.drawSegment(ctx, p1, p2);
      p1 = this._points[i];
      p2 = this._points[i + 1];
    }
    // Draw last line as a straight line while
    // we wait for the next point to be able to calculate
    // the bezier control point
    ctx.lineTo(p1.x, p1.y);
    ctx.stroke();
    ctx.restore();
  }

  /**
   * Converts points to SVG path
   * @param {Array} points Array of points
   * @return {PathData} SVG path commands
   */
  convertPointsToSVGPath(points: Point[]): PathData {
    const correction = this.width / 1000;
    return getSmoothPathFromPoints(points, correction);
  }

  /**
   * Creates a Path object to add on canvas
   * @param {PathData} pathData Path data
   * @return {Path} Path to add on canvas
   */
  createPath(pathData: PathData) {
    const path = new Path(pathData, {
      fill: null,
      stroke: this.color,
      strokeWidth: this.width,
      strokeLineCap: this.strokeLineCap,
      strokeMiterLimit: this.strokeMiterLimit,
      strokeLineJoin: this.strokeLineJoin,
      strokeDashArray: this.strokeDashArray,
    });
    if (this.shadow) {
      this.shadow.affectStroke = true;
      path.shadow = new Shadow(this.shadow);
    }

    return path;
  }

  /**
   * Decimate points array with the decimate value
   */
  decimatePoints(points: Point[], distance: number) {
    if (points.length <= 2) {
      return points;
    }
    let lastPoint = points[0],
      cDistance;
    const zoom = this.canvas.getZoom(),
      adjustedDistance = Math.pow(distance / zoom, 2),
      l = points.length - 1,
      newPoints = [lastPoint];
    for (let i = 1; i < l - 1; i++) {
      cDistance =
        Math.pow(lastPoint.x - points[i].x, 2) +
        Math.pow(lastPoint.y - points[i].y, 2);
      if (cDistance >= adjustedDistance) {
        lastPoint = points[i];
        newPoints.push(lastPoint);
      }
    }
    // Add the last point from the original line to the end of the array.
    // This ensures decimate doesn't delete the last point on the line, and ensures the line is > 1 point.
    newPoints.push(points[l]);
    return newPoints;
  }

  /**
   * On mouseup after drawing the path on contextTop canvas
   * we use the points captured to create an new Path object
   * and add it to the canvas.
   */
  _finalizeAndAddPath() {
    const ctx = this.canvas.contextTop;
    ctx.closePath();
    if (this.decimate) {
      this._points = this.decimatePoints(this._points, this.decimate);
    }
    const pathData = this.convertPointsToSVGPath(this._points);
    if (isEmptySVGPath(pathData)) {
      // do not create 0 width/height paths, as they are
      // rendered inconsistently across browsers
      // Firefox 4, for example, renders a dot,
      // whereas Chrome 10 renders nothing
      this.canvas.requestRenderAll();
      return;
    }

    const path = this.createPath(pathData);
    this.canvas.clearContext(this.canvas.contextTop);
    this.canvas.fire('before:path:created', { path: path });
    this.canvas.add(path);
    this.canvas.requestRenderAll();
    path.setCoords();
    this._resetShadow();

    // fire event 'path' created
    this.canvas.fire('path:created', { path: path });
  }
}
