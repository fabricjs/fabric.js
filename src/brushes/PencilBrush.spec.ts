import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { Canvas } from '../canvas/Canvas';
import { PencilBrush } from './PencilBrush';
import { parsePath } from '../util';
import type { TPointerEvent } from '../EventTypeDefs';
import { Path } from '../shapes/Path';
import { Point } from '../Point';

describe('PencilBrush', () => {
  let canvas: Canvas;

  beforeEach(() => {
    canvas = new Canvas();
  });

  afterEach(() => {
    canvas.cancelRequestedRender();
    canvas.off();
  });

  it('initializes constructor correctly', () => {
    expect(PencilBrush).toBeTruthy();
    const brush = new PencilBrush(canvas);
    expect(brush.canvas, 'assigns canvas').toBe(canvas);
    // @ts-expect-error -- protected
    expect(brush._points, 'points is an empty array').toEqual([]);
  });

  it('decimates points correctly', () => {
    const brush = new PencilBrush(canvas);
    const points = [
      new Point(1, 0),
      new Point(2, 0),
      new Point(3, 0),
      new Point(4, 0),
      new Point(5, 0),
    ];
    const distance = 6;
    const newPoints = brush.decimatePoints(points, distance);
    expect(newPoints[0], 'first point is always present').toBe(points[0]);
    expect(newPoints[1], 'last point is always present').toBe(
      points[points.length - 1],
    );
    expect(newPoints.length, 'All points removed except first and last').toBe(
      2,
    );
  });

  describe.each([true, false])(
    'with canvas.enableRetinaScaling = %s',
    (retinaScaling) => {
      beforeEach(() => {
        canvas.enableRetinaScaling = retinaScaling;
      });

      it('draws a point correctly', () => {
        const brush = new PencilBrush(canvas);
        const e = { target: canvas.upperCanvasEl } as unknown as TPointerEvent;
        const pointer = canvas.getScenePoint({
          ...e,
          clientX: 10,
          clientY: 10,
        });
        brush.onMouseDown(pointer, { e });
        // @ts-expect-error -- protected
        const pathData = brush.convertPointsToSVGPath(brush._points);
        expect(
          pathData,
          'path data create a small line that looks like a point',
        ).toEqual(parsePath('M 9.999 10 L 10.001 10'));
      });

      it('handles multiple coincident points', () => {
        const brush = new PencilBrush(canvas);
        const e = { target: canvas.upperCanvasEl } as unknown as TPointerEvent;
        const pointer = canvas.getScenePoint({
          ...e,
          clientX: 10,
          clientY: 10,
        });
        brush.onMouseDown(pointer, { e });
        brush.onMouseMove(pointer, { e });
        brush.onMouseMove(pointer, { e });
        brush.onMouseMove(pointer, { e });
        brush.onMouseMove(pointer, { e });
        // @ts-expect-error -- protected
        const pathData = brush.convertPointsToSVGPath(brush._points);
        expect(
          pathData,
          'path data create a small line that looks like a point',
        ).toEqual(parsePath('M 9.999 10 L 10.001 10'));
        // @ts-expect-error -- protected
        expect(brush._points.length, 'concident points are discarded').toBe(2);
      });

      it('handles multiple non-coincident points', () => {
        const brush = new PencilBrush(canvas);
        const e = { target: canvas.upperCanvasEl } as unknown as TPointerEvent;
        const pointer = canvas.getScenePoint({
          ...e,
          clientX: 10,
          clientY: 10,
        });
        const pointer2 = canvas.getScenePoint({
          ...e,
          clientX: 15,
          clientY: 15,
        });
        const pointer3 = canvas.getScenePoint({
          ...e,
          clientX: 20,
          clientY: 20,
        });
        brush.onMouseDown(pointer, { e });
        brush.onMouseMove(pointer2, { e });
        brush.onMouseMove(pointer3, { e });
        brush.onMouseMove(pointer2, { e });
        brush.onMouseMove(pointer3, { e });
        // @ts-expect-error -- protected
        const pathData = brush.convertPointsToSVGPath(brush._points);
        expect(pathData, 'path data create a complex path').toEqual(
          parsePath(
            'M 9.999 9.999 Q 10 10 12.5 12.5 Q 15 15 17.5 17.5 Q 20 20 17.5 17.5 Q 15 15 17.5 17.5 L 20.001 20.001',
          ),
        );
        // @ts-expect-error -- protected
        expect(brush._points.length, 'concident points are discarded').toBe(6);
      });

      it('handles points outside canvas', () => {
        const brush = new PencilBrush(canvas);
        const e = { target: canvas.upperCanvasEl } as unknown as TPointerEvent;
        const pointer = canvas.getScenePoint({
          ...e,
          clientX: 10,
          clientY: 10,
        });
        const pointer2 = canvas.getScenePoint({
          ...e,
          clientX: 15,
          clientY: 100,
        });
        const pointer3 = canvas.getScenePoint({
          ...e,
          clientX: 20,
          clientY: 160,
        });
        const pointer4 = canvas.getScenePoint({
          ...e,
          clientX: 320,
          clientY: 100,
        });
        const pointer5 = canvas.getScenePoint({
          ...e,
          clientX: 100,
          clientY: 100,
        });
        brush.onMouseDown(pointer, { e });
        brush.onMouseMove(pointer2, { e });
        brush.onMouseMove(pointer3, { e });
        brush.onMouseMove(pointer4, { e });
        brush.onMouseMove(pointer5, { e });
        // @ts-expect-error -- protected
        const pathData = brush.convertPointsToSVGPath(brush._points);
        expect(
          pathData,
          'path data create a path that goes beyond canvas',
        ).toEqual(
          parsePath(
            'M 9.999 9.999 Q 10 10 12.5 55 Q 15 100 17.5 130 Q 20 160 170 130 Q 320 100 210 100 L 99.999 100',
          ),
        );
        // @ts-expect-error -- protected
        expect(brush._points.length, 'all points are available').toBe(6);
      });

      it('limits points to canvas size when limitedToCanvasSize is true', () => {
        const brush = new PencilBrush(canvas);
        brush.limitedToCanvasSize = true;
        const e = { target: canvas.upperCanvasEl } as unknown as TPointerEvent;
        const pointer = canvas.getScenePoint({
          ...e,
          clientX: 10,
          clientY: 10,
        });
        const pointer2 = canvas.getScenePoint({
          ...e,
          clientX: 15,
          clientY: 100,
        });
        const pointer3 = canvas.getScenePoint({
          ...e,
          clientX: 20,
          clientY: 160,
        });
        const pointer4 = canvas.getScenePoint({
          ...e,
          clientX: 320,
          clientY: 100,
        });
        const pointer5 = canvas.getScenePoint({
          ...e,
          clientX: 100,
          clientY: 100,
        });
        brush.onMouseDown(pointer, { e });
        brush.onMouseMove(pointer2, { e });
        brush.onMouseMove(pointer3, { e });
        brush.onMouseMove(pointer4, { e });
        brush.onMouseMove(pointer5, { e });
        // @ts-expect-error -- protected
        const pathData = brush.convertPointsToSVGPath(brush._points);
        expect(
          pathData,
          'path data create a path that does not go beyond canvas',
        ).toEqual(
          parsePath(
            'M 9.999 9.999 Q 10 10 12.5 55 Q 15 100 57.5 100 L 100.001 100',
          ),
        );
        // @ts-expect-error -- protected
        expect(brush._points.length, '2 points have been discarded').toBe(4);
      });

      it('fires events and creates path on mouse up', () => {
        let fireBeforePathCreatedEvent = false;
        let firePathCreatedEvent = false;
        let added = null;

        canvas.on('before:path:created', () => {
          fireBeforePathCreatedEvent = true;
        });

        canvas.on('path:created', (opt) => {
          firePathCreatedEvent = true;
          added = opt.path;
        });

        const brush = new PencilBrush(canvas);
        const e = { target: canvas.upperCanvasEl } as unknown as TPointerEvent;
        const pointer = canvas.getScenePoint({
          ...e,
          clientX: 10,
          clientY: 10,
        });
        const pointer2 = canvas.getScenePoint({
          ...e,
          clientX: 15,
          clientY: 15,
        });
        const pointer3 = canvas.getScenePoint({
          ...e,
          clientX: 20,
          clientY: 20,
        });

        brush.onMouseDown(pointer, { e });
        brush.onMouseMove(pointer2, { e });
        brush.onMouseMove(pointer3, { e });
        brush.onMouseMove(pointer2, { e });
        brush.onMouseMove(pointer3, { e });
        brush.onMouseUp({ e });

        expect(
          fireBeforePathCreatedEvent,
          'before:path:created event is fired',
        ).toBe(true);
        expect(firePathCreatedEvent, 'path:created event is fired').toBe(true);
        expect(added, 'a path is added').toBeInstanceOf(Path);
        expect(added!.path).toMatchSnapshot();
        expect(added!.path.length, 'path has 6 steps').toBe(4);
      });
    },
  );
});
