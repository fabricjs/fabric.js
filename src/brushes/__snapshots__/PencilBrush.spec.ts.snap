// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PencilBrush > with canvas.enableRetinaScaling = false > fires events and creates path on mouse up 1`] = `
[
  [
    "M",
    9.999,
    9.999,
  ],
  [
    "Q",
    10,
    10,
    12.5,
    12.5,
  ],
  [
    "Q",
    15,
    15,
    17.5,
    17.5,
  ],
  [
    "L",
    20,
    20,
  ],
]
`;

exports[`PencilBrush > with canvas.enableRetinaScaling = true > fires events and creates path on mouse up 1`] = `
[
  [
    "M",
    9.999,
    9.999,
  ],
  [
    "Q",
    10,
    10,
    12.5,
    12.5,
  ],
  [
    "Q",
    15,
    15,
    17.5,
    17.5,
  ],
  [
    "L",
    20,
    20,
  ],
]
`;
